### ISSUE TITLE
Upgrade to WebKit2

### ISSUE THREAD

**Issue by mvz** [lwindolf/liferea#311]
WebKit 1 has security issues that won't be resolved. Please consider upgrading to WebKit2.

See https://blogs.gnome.org/mcatanzaro/2016/02/01/on-webkit-security-updates/

**Comment by Leiaz** [comment]
I've looked into this (see [webkit2 branch.](https://github.com/Leiaz/liferea/tree/webkit2)) Upgrading would fix #315.

Some things are broken :

WebKit 2 uses global proxy settings, there is no API to set a custom proxy : 
- https://bugs.webkit.org/show_bug.cgi?id=128674
- https://bugs.webkit.org/show_bug.cgi?id=113663

WebKitWebView makes its own scrollbars and doesn't implement GtkScrollable anymore. The scrollbars are not in the same style as other scrollbars in the application. And the headlines skimming functionality (Ctrl+Space) is broken. 

The functions to check if something is selected are gone, so the "Copy" action in the context menu is always active.

**Comment by lwindolf** [comment]
@Leiaz The feature breaking make this a hard upgrade. I could live with the "Copy" action missing, but the skimming is not nice.

**Comment by Leiaz** [comment]
It can probably be made to work by making a [Web Extension for WebKit](https://blogs.igalia.com/carlosgc/2013/09/10/webkit2gtk-web-process-extensions/) and using the DOM API.

Epiphany can serve [as example](https://git.gnome.org/browse/epiphany/tree/embed/web-extension/), they use DBus to communicate with the extension.

**Comment by lwindolf** [comment]
Argh... this cannot be right, everyone using Webkit2Gtk needs to do this? Then it belongs into Webkit2Gtk...

**Comment by bitlord** [comment]
> The scrollbars are not in the same style as other scrollbars in the application.  
I guess this is fixed with latest >=webkit-gtk-2.11.90[1]

- Switch to use overlay scrollbars like all other GTK+ widgets and ensure the behavior is consistent with GTK+ too.


[1] [http://webkitgtk.org/2016/02/19/webkitgtk2.11.90-released.html](http://webkitgtk.org/2016/02/19/webkitgtk2.11.90-released.html )

**Comment by rich-coe** [comment]
@Leiaz  I tried buiding this, but I suspect I'm doing something wrong as I get blank window where the feed item is displayed.  Is there a minimum webkit2-gtk needed for this to work?

**Comment by rich-coe** [comment]
I'm trying out webkit2 patch
(1) when I view the slashdot rss feed I get this error:
Received an invalid message "WebPageProxy.DidCommitLoadForFrame" from the web process.
And then the item view entry goes blank on redraw.

(2) when I open an item in a new tab or visit a page in a new tab I get this error
Received an invalid message "WebProcessProxy.AddBackForwardItem" from the web process.

I could not find where these messages are coming from.  A web search for (2) found a webkit bug which claimed the bug writer was mixing old webkit code with new webkit code.

**Comment by Leiaz** [comment]
I'm using webkit2gtk 2.10.7, but 2.6 should be enough. I'm not seeing those errors. Which version are you using ?

**Comment by rich-coe** [comment]
I had webkit2gtk 2.6.6 installed, but after updating to 2.10.7, still had issues because the dependencies had not installed.  I also had issues, because I had older versions of some webkit/pango dependencies installed in /usr/local.

**Comment by rich-coe** [comment]
Some observations on the webkit2 port.
(1) Control-space no longer scrolls the current article, but acts like 'next' when the article is complete.
(2) If the displayed page crashes, the liferea app doesn't crash.
(3) I have not had any javascript errors.  Still testing.  So far so good.

**Comment by rich-coe** [comment]
Control-space is working. Thank you.

I noticed this message before, during startup and when a picture paints, and didn't log it before.
(WebKitWebProcess:24372): GLib-GObject-WARNING **: gsignal.c:2516: signal 'starting' is invalid for instance '0x21659b0' of type 'SoupMessage'

**Comment by Leiaz** [comment]
@rich-coe  This is from WebKit, [the doc here](https://developer.gnome.org/libsoup/stable/SoupMessage.html#SoupMessage-starting) says this signal is new to libsoup 2.50.

I've made some changes to do the menu with actions and use asynchronous calls.
The web extension is now only used for headlines skimming. It could be done in Javascript instead but this would require having Javascript enabled ...

I've removed the proxy settings. I wanted to change the deprecated GtkVBox GtkHBox to GtkBox but then I saw the comment [in the doc](https://developer.gnome.org/gtk3/stable/GtkVBox.html) saying that GtkGrid should be used because GtkBox is going to be deprecated too. So I didn't do it but it may be simpler to remake the ui files in the new Glade.

@lwindolf Do I send a pull request or do you prefer Javascript to a web extension ?

**Comment by mcatanzaro** [comment]
Note that we only connect to SoupMessage::starting if compiled against sufficiently-new libsoup. The only way you could be seeing that error is if your WebKit was compiled against a newer version of libsoup than you have installed (a serious distro bug).

It was added here: http://trac.webkit.org/changeset/180928

**Comment by rich-coe** [comment]
@mcatanzaro  I upgraded libsoup and that fixed it.  I had upgraded webkit from a new repository.  Unfortunately rpm is not very good at dependencies.

**Comment by bitlord** [comment]
For some time I'm trying to make this run on my system, but it is always stuck at startup, not sure if this is a relevant part, but it may be, it tries to "move"/"rename" a file which doesn't exist. I tried to remove all confings/cache ... ~/.cache/liferea ~/.local/share/liferea ~/.config/liferea

`close(30)                               = 0
munmap(0x7fc828000000, 4096)            = 0
rename("/home/branko/.config/liferea/feedlist.opml~", "/home/branko/.config/liferea/feedlist.opml") = 0
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 2994) = 0 (Timeout)
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 9997) = 0 (Timeout)
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 9879) = 0 (Timeout)
close(34)                               = 0
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 109) = 0 (Timeout)
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 1026) = 0 (Timeout)
close(29)                               = 0
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 87) = 0 (Timeout)
close(28)                               = 0
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 328) = 0 (Timeout)
close(35)                               = 0
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 256) = 0 (Timeout)
close(27)                               = 0
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 8300) = 0 (Timeout)
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=5, events=POLLIN}, {fd=10, events=POLLIN}, {fd=22, events=POLLIN}], 4, 9994q) = 1 ([{fd=5, revents=POLLIN}])
recvmsg(5, {msg_name(0)=NULL, msg_iov(1)=[{"U\2\f\1=~U\0\3\24\4\0\20\0\0\0\0\0\0\24\24\24\24\24\0\0\3\37%\2\0\0", 4096}], msg_controllen=0, msg_flags=0}, 0) = 32
recvmsg(5, 0x7ffdd5ab45d0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(5, 0x7ffdd5ab45b0, 0)           = -1 EAGAIN (Resource temporarily unavailable)
`

**Comment by Leiaz** [comment]
The file `~/.config/liferea/feedlist.opml~` is only supposed to exist temporarily and this line :
``` 
rename("/home/branko/.config/liferea/feedlist.opml~", "/home/branko/.config/liferea/feedlist.opml") = 0
```
Tells you the return value was 0, which means it succeeded.

One reason Liferea may fail to start is if you didn't install it, it can't find the correct GSettings schema. Otherwise can you post the gdb backtrace ?

**Comment by bitlord** [comment]
Thanks @Leiaz, it is built locally and installed locally, I used --prefix=/home/branko/bin and I installed it locally. As for Gsettings schemas, I did have issues with them, but I found a solution 
GSETTINGS_SCHEMA_DIR="/home/branko/bin/share/glib-2.0/schemas/"
without it it "crashes" (Trap) at startup.

As for gdb, I don't have debug symbols, maybe I can rebuild some things and have debug symbols

`#0  0x00007fece1db51cd in poll () from /lib64/libc.so.6
#1  0x00007fece22d6eec in ?? () from /usr/lib64/libglib-2.0.so.0
#2  0x00007fece22d6ffc in g_main_context_iteration () from /usr/lib64/libglib-2.0.so.0
#3  0x00007fece2af6c0c in g_application_run () from /usr/lib64/libgio-2.0.so.0
#4  0x000000000041a856 in main (argc=1, argv=0x7ffcc1fa0808) at main.c:258
`

**Comment by Leiaz** [comment]
Liferea symbols should be enough for most things.

But I just realized, you mean it doesn't actually crash ? If you were using 1.10 before, [the tray icon is now a plugin, so your window may be hidden](https://lzone.de/liferea/blog/Why%20doesnt%20Liferea%20start%20anymore).

**Comment by bitlord** [comment]
Ok, that was the issue, I was thinking about it before, but I excluded it as a problem when I removed my current config.

It works now with '-w shown'.

Thank you @Leiaz.

**Comment by bitlord** [comment]
I must say that his port works really good for me. There is only one minor issue, sometimes when you start liferea (in my case it's in systray) and update feeds (maybe not important) try to open one news item there is nothing in the webview (or however it is called), just blank, as a workaround I change view mode (Normal, Wide, Combined), switch from one to another (it starts working), and back to preferred one "Normal". I'm still on webkit 2.10.x and gtk+ 3.18.x, I'll try updating webkit to 2.12. I guess it will work with other "older" libs...

**Comment by bitlord** [comment]
After upgrading to latest stable webkit, gtk+, ... it is much better experience.

I would love to see this "merged" into official repos, and being updated, it works very well, and stable for me. Not to mention advantages of using modern webkit with bug and security fixes. 

Thank you @Leiaz for this port, and thanks to @lwindolf and all contributors for liferea.

**Comment by lwindolf** [comment]
@bitlord Thanks for the feedback on your tests!

@Leiaz What is currently hindering me to merge it for 1.12 is the missing Webkit2 proxy support. Because it would hurt all non-GNOME users relying on Liferea's custom proxy settings.

**Comment by genodeftest** [comment]
Hm, when choosing between a rarely used feature and getting a bunch of known security bugs fixed I'd choose the latter.

**Comment by jbicha** [comment]
I believe it would be best if proxy settings were set system-wide. Should every app that uses the Internet have its own proxy settings dialog? Because I think that's probably a really bad experience for users.

**Comment by mcatanzaro** [comment]
For the record, the current situation:

 * glib-networking uses GNOME system settings (GSettings via GProxyResolverGNOME) if GNOME_DESKTOP_SESSION_ID is present in the environment; this should be sufficient for all GNOME users unless you want to use a different proxy just for Liferea for some reason.
 * Otherwise, glib-networking uses libproxy (via GLibProxyResolver) which just checks the FTP_PROXY, HTTPS_PROXY, and HTTP_PROXY environment variables.

So the system-wide configuration (which is definitely best) should already handled satisfactorily.

I am not sure why anyone would want to use different proxy settings just for Liferea -- we would certainly never expose such settings in a GNOME app, for the reason @jbicha suggests -- but my understanding is that's the only missing feature here. So what exactly is needed from WebKit? Just a way to set the proxy-uri property of the SoupSession in the network process? If so, that shouldn't be too hard, it's just a matter of someone finding time to work on it and get a patch adding that new API through the WebKit review process. But if you want to be able to use a custom GProxyResolver, that's not going to be technically possible unfortunately.

**Comment by genodeftest** [comment]
@mcatanzaro : When using e.g. Liferea with Tor, setting proxy through GNOME or through environment variables is a bad idea if you also want to use web services which will leak private data (e.g. gnome weather app (by design) or any browser on some websites or gnome-contacts with [this](https://bugzilla.gnome.org/show_bug.cgi?id=750192) bug). Users might want to separate those pieces of information from their RSS feeds.
Also, many provies (including Tor), are known to attack users. Using RSS through them would be ok, but not something like evolution or your web mail account.

**Comment by lwindolf** [comment]
@mcatanzaro Another use case are the non-GNOME users (Windows, XFCE, Enlightenment). Dropping the manual proxy setting would remove usability a lot for those.

**Comment by lwindolf** [comment]
@jbicha I agree the UX is not perfect. Still Liferea is not a GNOME application and cannot rely on the GNOME defaults. We workaround the UX issues by closely following the major dialog layouts (GNOME and Firefox) while defaulting to the GNOME settings.

**Comment by lwindolf** [comment]
I believe we need to follow the environment variable hack to workaround like implemented by Nuvola (https://github.com/tiliado/nuvolaplayer/blob/8329274861a0d9258d6ea606c351633d5798440f/src/nuvolakit-runner/system/Connection.vala) as suggested in (https://bugs.webkit.org/show_bug.cgi?id=128674). It should have low impact on complexity and is reasonable to test. 

This way we can switch to recent Webkit.

@Leiaz I'll merge your implementation soon. I hope you do not mind.

**Comment by mcatanzaro** [comment]
Nuvola's hack doesn't look safe, we've learned from crashes in GNOME that you can't modify the environment in a threaded program after creating your first secondary thread, as it's impractical to ensure the other threads will never access it at the same time. (setenv needs to be done at the very top of main, or it's almost surely not safe.) It'd be hard to do with a preferences dialog anyway. I think we just need to add some API along the lines of `webkit_web_context_set_proxy_uri` to WebKit.

**Comment by Leiaz** [comment]
@mcatanzaro Yes it would be enough, setting the proxy-uri of the SoupSession is how it is done with WebKit 1.

@lwindolf I had removed the proxy settings, as the environment variable thing is a bit hacky and the users who want a different proxy for Liferea could set the env variables themselves (but yes it is less user friendly than a GUI). I think setting it the way Nuvola does will require restarting Liferea for changes to take effect.

**Comment by rich-coe** [comment]
The youtube video at embedded in this web page
https://www.geeksaresexy.net/2016/07/27/internet-ads-salesmen/
kills the webkit code.  
I think it's a webkit issue, but I haven't tried it yet with the standalone webkit viewer.
webkit2gtk (4.0 I believe)

**Comment by mcatanzaro** [comment]
I have a WIP patch for WebKit in https://bugs.webkit.org/show_bug.cgi?id=128674, will try to finish it soonish.

**Comment by mcatanzaro** [comment]
I updated the WebKit bug with a new patch. Please test it out and see if it's sufficient for Liferea.

**Comment by mcatanzaro** [comment]
Hm. We could add a way to guarantee no proxy will be used (perhaps by passing "direct://" as the proxy URI?) That seems like it might be useful to have and easy to implement; what do you think? (It would unset both the proxy-uri and proxy-resolver properties of the SoupSession.)

Although, I wonder if, considering the fact that nobody noticed the existing Liferea option was broken until now, indicates that maybe this setting is not so important.

**Comment by Leiaz** [comment]
With the api proposed on the bug page it shouldn't be a problem to (fix and) keep the option.

**Comment by genodeftest** [comment]
The "Scrollbars have different style in WebKit2" issue is tracked here: https://bugs.webkit.org/show_bug.cgi?id=161735

**Comment by mcatanzaro** [comment]
Note that we didn't know about the scrollbar problem until today, since nobody ever reported it and we only test Adwaita.

**Comment by Leiaz** [comment]
The scrollbars work fine here since 2.12 (or around that). They just weren't styled before. That bug is different.

**Comment by rich-coe** [comment]
On KDE Plasma 5, there's some UI widget that overlays the gtk scrollbar.  I'm not sure why or what problem it's attempting to fix.  It's ugly also.

**Comment by mcatanzaro** [comment]
Does it look better in KDE if you run with GTK_OVERLAY_SCROLLING=0?

If other GTK+ 3 apps do not experience this issue, please file a new WebKit bug separate from Milan's.

**Comment by rich-coe** [comment]
@mcatanzaro GTK_OVERLAY_SCROLLING=0  just makes the ugly scoll bar appear all the time instead of auto-hiding.

**Comment by lwindolf** [comment]
I've merged @Leiaz Webkit2 implementation. @Leiaz I marked all as GPLv2+ with you as author. Can you give me an "OK" for this please?

@all: I do not have scroll bar issues in GNOME and Unity, but will continue watching out for those.

Issues that need fixing before 1.12 though:
* Article skimming broke
* Very small font size on my system (maybe a HiDPI problem)
* a Webkit2 version dependant use of the new proxy (#ifdef version > x.y and stdout info message otherwise)

Nonetheless I will close this ticket as it is about the Webkit2 migration and track the other issues as 1.12 integration tickets. Please everyone start posting tickets on

**Comment by Leiaz** [comment]
@lwindolf  OK. Another possible issue : I noticed that I'm using [Glib macros](https://developer.gnome.org/gobject/stable/gobject-Type-Information.html#G-DECLARE-FINAL-TYPE:CAPS) that avoid a lot of boilerplate for GTypes but are new to glib 2.44. Current Debian stable (8 Jessie) is only at 2.42, but it has webkit2gtk-4.0. Do I make it work on Debian stable ?

For article skimming I'm afraid that `webkit_dom_dom_window_scroll_by` could be asynchronous ...

**Comment by mcatanzaro** [comment]
Regarding small fonts: the font size API is terrible to use and needs to be replaced, but in the meantime you have to handle it manually to get hidpi right. There is some boilerplate code you can copypaste from epiphany for this. See https://bugs.webkit.org/show_bug.cgi?id=142673 and https://git.gnome.org/browse/epiphany/tree/embed/ephy-embed-prefs.c#n206

Regarding proxy support: I'll try to have a new WebKit patch ready in the next week or two; my previous API proposal was rejected.

**Comment by mcatanzaro** [comment]
The proxy API has landed in WebKit; it will be available in WebKitGTK+ 2.16. Better late than never!

**Comment by lwindolf** [comment]
Great! Thanks for letting us now @mcatanzaro !
